### PR TITLE
Add support for platform checkout server side payment processing

### PR DESCRIPTION
**Pull request by reykjalin**
## Changes proposed in this Pull Request

Supersedes #3642 

This PR adds the necessary functionality for us to support platform payment processing without introducing all the new filters we needed for the first approach we tried in #3642 .

- Add a new filter, `wc_payments_account_id_for_intent_confirmation`, to allow the platform to modify the account used to confirm a payment intent.
    - **Note:** The account ID must match the provided payment intent, otherwise the intent confirmation will fail.
- Add new JS prop to the frontend settings, `accountIdForIntentConfirmation`, which contains the result of running the new filter on an empty string.
- Add the ability to process a payment when the checkout request provides a payment intent.
    - This is so the platform can process the payment and then update the order on the merchant store appropriately.

## Testing instructions

1. Make sure payments work as they normally do in a normal WCPay store. Maybe try some of the [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows)?
2. Follow the testing instructions in the corresponding platform checkout PR (should be mentioned below via GitHub mention).

I have no automated tests for this because I'm not so sure how to actually cover this with automated tests.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : QA Testing Not Applicable

### CODE REVIEW

File: includes/class-wc-payment-gateway-wcpay.php
Comment (reykjalin) [inline includes/class-wc-payment-gateway-wcpay.php:2206]: See https://github.com/Automattic/woocommerce-payments/pull/3642/files#r786451779

**Comment by reykjalin** [comment]
I'm not so sure this needs a changelog entry. I think the filter should at least have an entry, but I'm not sure about the other stuff.

Would love to hear from the reviewer on that.

**Review by asumaran — APPROVED** [review]
LGTM

**Comment by reykjalin** [comment]
@ricardo It's been a good while since the last review, so pinging you here for a review as well.

File: client/checkout/api/index.js
Comment (ricardo) [inline client/checkout/api/index.js:259]: This part is kinda tricky IMO, and even more now with the extra condition for the different account ID. I wonder if we can make it easier to understand what's going on.

Some ideas came to mind:

- Perhaps we should call the `getStripe` function here something else, to help differentiate between this and `WCPayAPI.getStripe`?
- Or, does it make sense to create a function to return the intent confirmation (confirmCardSetup/confirmCardPayment)? We can at least dismiss the extra `isSetupIntent` check.

For example:

```JavaScript
const confirmIntention = () => {
	const { locale, publishableKey } = this.options;
	const accountIdForIntentConfirmation = getConfig(
		'accountIdForIntentConfirmation'
	);

	// If this is a setup intent we're not processing a platform checkout payment so we can
	// use the regular getStripe function.
	if ( isSetupIntent ) {
		return this.getStripe().confirmCardSetup( clientSecret );
	}

	if ( accountIdForIntentConfirmation ) {
		return this.createStripe(
			publishableKey,
			locale,
			accountIdForIntentConfirmation
		).confirmCardPayment( clientSecret );
	}

	// When not dealing with a setup intent or platform checkout we need to force an account
	// specific request in Stripe.
	return this.getStripe( true ).confirmCardPayment( clientSecret );
};

const request = confirmIntention()
```

- Or even, avoid this altogether by overriding `options.accountId` when initializing the WCPay API with `getConfig( 'accountIdForIntentConfirmation' )`.

```JavaScript
// Or could this be overwritten from the back-end instead of adding "accountIdForIntentConfirmation"?
const accountId = getConfig( 'accountIdForIntentConfirmation' ) ? getConfig( 'accountIdForIntentConfirmation' ) : getConfig( 'accountId' );

new WCPayAPI(
	{
		publishableKey,
		accountId,
		forceNetworkSavedCards: getConfig( 'forceNetworkSavedCards' ),
		locale: getConfig( 'locale' ),
	}
```

And we can keep this part of the code as it was before.

```JavaScript
const confirmAction = isSetupIntent
	? this.getStripe().confirmCardSetup( clientSecret )
	: this.getStripe( true ).confirmCardPayment( clientSecret );
```

PS: Haven't tested the code here though 😅.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (ricardo) [inline includes/class-wc-payment-gateway-wcpay.php:636]: Should the platform checkout always initialize Stripe as the platform, except when confirming an intent?

If so, perhaps we could overwrite the `accountId` instead. That would prevent a lot of maybe unnecessary JS changes I mentioned in the other comment.

**Review by ricardo — CHANGES_REQUESTED** [review]

File: includes/class-wc-payment-gateway-wcpay.php
Comment (reykjalin) [reply includes/class-wc-payment-gateway-wcpay.php:629]: Ah, this was accidental 🤔 
I only meant to add the `accountIdForIntentConfirmation` prop, but something must've messed things up during a merge/rebase.

I'll go through all the options and fix this 👍 

Thanks for catching this!

File: includes/class-wc-payment-gateway-wcpay.php
Comment (reykjalin) [reply includes/class-wc-payment-gateway-wcpay.php:636]: > Should the platform checkout always initialize Stripe as the platform, except when confirming an intent?

On the platform checkout site, _yes_. In other words; on the platform checkout site we want to use the platform account (i.e. the store account) for all cases, _except_ when confirming an intent. The reason for that is that the payment intent is created with the merchant account, so we need the same account to confirm the intent.

> If so, perhaps we could overwrite the `accountId` instead

That was the initial approach, but we decided not to go that route to limit the impact of the new filter as much as possible. If we overwrite the `accountId` that would give plugin authors (or anyone with write access to the site) the ability to overwrite the account ID used in the Stripe SDK. By adding a new property we can ensure that the value from the filter is only used when confirming intents.

It's inconvenient and complicates the JS, but provides better encapsulation for the value returned by the filter.

File: client/checkout/api/index.js
Comment (reykjalin) [reply client/checkout/api/index.js:259]: > Perhaps we should call the `getStripe` function here something else, to help differentiate between this and `WCPayAPI.getStripe?`

I did call them the same thing because they sort of serve the same purpose, but I can definitely see how that could be confusing 😅 

> avoid this altogether by overriding `options.accountId` when initializing the WCPay API with `getConfig( 'accountIdForIntentConfirmation' )`.

Unfortunately this won't work since the `WCPayAPI.getStripe()` function "caches" the Stripe SDK instance it creates, so we'll have to recreate the Stripe SDK instance here anyway. The reason being that the SDK should normally be instantiated with the store account ID, but in some cases (e.g. platform checkout intent confirmation) we need a different account ID, so we'll need 2 Stripe JS SDKs initialized at this point.

> does it make sense to create a function to return the intent confirmation (confirmCardSetup/confirmCardPayment)?

Hmm, I like this idea 🤔 

Let's go with that and see if that improves things 👏

File: includes/class-wc-payment-gateway-wcpay.php
Comment (reykjalin) [reply includes/class-wc-payment-gateway-wcpay.php]: Fixed in e7a135899cb6ef7083ed267378e6decb7a0da9bf 🙌

File: includes/class-wc-payment-gateway-wcpay.php
Comment (ricardo) [reply includes/class-wc-payment-gateway-wcpay.php:636]: > If we overwrite the accountId that would give plugin authors (or anyone with write access to the site) the ability to overwrite the account ID used in the Stripe SDK.

Got it. Makes sense.

File: client/checkout/api/index.js
Comment (reykjalin) [reply client/checkout/api/index.js]: I tried this out in https://github.com/Automattic/woocommerce-payments/pull/3811/commits/c245507af26b1e264a92da94bd5f4e2e74981a5c, but not sure if it looks better 🤔 
What do you think?

File: includes/class-wc-payment-gateway-wcpay.php
Comment (ricardo) [reply includes/class-wc-payment-gateway-wcpay.php:1106]: Yeah, maybe `sanitize_key` 👍.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (ricardo) [reply includes/class-wc-payment-gateway-wcpay.php:2200]: Should we try to confirm or fix this bug before merging the PR?

**Review by ricardo — APPROVED** [review]
LGTM 👍

> I tried this out in https://github.com/Automattic/woocommerce-payments/commit/c245507af26b1e264a92da94bd5f4e2e74981a5c, but not sure if it looks better 🤔 What do you think?

I think it got a bit clearer 👍. It was mainly confusing to me because of the function name, and now we have a bit less code to process as well.

File: includes/class-wc-payment-gateway-wcpay.php
Comment (reykjalin) [reply includes/class-wc-payment-gateway-wcpay.php:2200]: I'll see if it's still happening before merging and report back 🤔

File: includes/class-wc-payment-gateway-wcpay.php
Comment (reykjalin) [reply includes/class-wc-payment-gateway-wcpay.php]: Ah, turns out sanitize_key won't work since it filters out upper case characters and Stripe IDs may have upper case characters 😩

I think [sanitize_user( $intent_id, true /* enable strict mode */ )](https://developer.wordpress.org/reference/functions/sanitize_user/) is more appropriate since it will not replace upper case characters 🤔 

Made that change in df263b3e96c4c6f078a16a5ac52c1aaf31b71937

File: includes/class-wc-payment-gateway-wcpay.php
Comment (reykjalin) [reply includes/class-wc-payment-gateway-wcpay.php:2202]: Just confirmed that this is still an issue. Very easy to reproduce;

Just run a payment. If you set a breakpoint here and inspect the value you'll see that it comes back as `"null"` 🤷‍♂️

**Pull request merged**
